1주차

mindock opened the pull request:
## 구현 내용
- sku, center, barcode 도메인 정의
- center, sku 생성/수정/상세조회 API 구현

mindock committed:
� sku, center, barcode �¦硫鯛 異�

mindock committed:
� h2 �ㅼ� 異�

mindock committed:
� center �庖¦/��/�¬�몄“� API 異�

mindock committed:
� sku �庖¦/��/�¬�몄“� API 異�

vsh123 commented on build.gradle.kts:
```diff
@@ -5,8 +5,14 @@ plugins {
+    kotlin("plugin.jpa") version "1.6.21"
 }
 
+allOpen {
```
크 Lazy를 위한 allOpen인거죠? 👍

mindock replied:
맞습니당~!

vsh123 commented on src/main/kotlin/com/mindock/minifullfillment/controller/CenterController.kt:
```diff
@@ -0,0 +1,42 @@
+import org.springframework.web.bind.annotation.*
+
+@RestController
+@RequestMapping("/api/v1/centers")
```
api versioning 👍

vsh123 commented on src/main/kotlin/com/mindock/minifullfillment/controller/CenterController.kt:
```diff
@@ -0,0 +1,42 @@
+) {
+
+    @PostMapping
+    fun create(
+        @RequestBody request: CenterCreateRequest
+    ): CenterResponse {
```
얘네는 한 라인에 만드는 것도 나쁘지 않아보이는데, 어떻게 생각하시나용

mindock replied:
1개일때는 한줄로 만드는 것도 좋은 것 같아요~
제 개인적인 생각은 2개 이상일 때부터 줄바꿈하는 것이 좋을 것 같아요~!

vsh123 replied:
좋습니다아~~~~~!!

vsh123 commented on src/main/kotlin/com/mindock/minifullfillment/domain/Sku.kt:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    companion object {
+        fun of(name: String, code: String): Sku =
```
READY상태의 SKU를 만들어주는 팩터리 메서드라면 조금 더 네이밍을 명확하게 해주면 어떨까여?
예를들어 READY가 아닌 SALE을 만들어주는 팩터리 메서드가 추가된다면?? 해다 함수는 네이밍을 어떻게 지어야 할까?

mindock replied:
요 부분은 of 메소드에 status를 받아서 설정하는 것으로 수정할게요~!
말씀대로 READY 상태로 항상 만들어준다면 메소드명을 의미있는 네이밍으로 변경해야할 것 같네요~!

vsh123 commented on src/main/kotlin/com/mindock/minifullfillment/domain/Stock.kt:
```diff
@@ -0,0 +1,22 @@
+package com.mindock.minifullfillment.domain
+
+import javax.persistence.*
```
나는 ktlint를 사용하는데, 린트에서는 기본적으로 와일드카드 import를 금지시키구 있어!!
너네는 어때?

[참고](https://blog.leocat.kr/notes/2020/12/14/intellij-avoid-wildcard-imports-in-kotlin-with-intellij)

mindock replied:
오 회사에서는 ktlint 적용 시도하다가(적용은 못함😵‍💫) 해당 설정 적용되어있는데, 요 작업은 개인 노트북이라 설정을 안했네용~!
해당 설정 추가할게요~

vsh123 commented on src/main/kotlin/com/mindock/minifullfillment/domain/Stock.kt:
```diff
@@ -0,0 +1,22 @@
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "stock_id")
```
PK의 컬럼명을 단순 id라고 표시하는게 아닌 entity_id 라고 명시해주는 이유가 궁금합니당

mindock replied:
회사에서 사용하는 컨벤션이라 사용했던 것 같아요..ㅎㅎㅎ
entity_id, id 어떤것을 사용해도 괜찮을 것 같은데 혹시 @vsh123 님 생각은 어떤가여?

vsh123 replied:
저는 개인적으로 여기서 명시된 id는 PK 이상, 이하의 의미도 가지지 않는다구 생각해요!!
joinColumn이라면 stock_id와 같이 도메인의 id라고 명시를 해주겠지만, 이경우에는 그냥 id로만 해도 괜찮다구 생각해요~!

vsh123 commented on src/main/kotlin/com/mindock/minifullfillment/domain/Stock.kt:
```diff
@@ -0,0 +1,22 @@
+    @JoinColumn(name = "sku_id")
+    val sku: Sku,
+
+    var quantity: Int
```
* 이부분은 내가 코틀린하면서 가장고민하는점인데, 외부에 setter를 노출하는 것과 같은 효과이자너?? 이런부분은 어떻게 생각해?
* quantity는 적어도 Long타입으로 선언해주면 어떨까?

mindock replied:
> quantity는 적어도 Long타입으로 선언해주면 어떨까?

재고 수량이 Long 타입까지 크게 들어오나요??

vsh123 replied:
사실 서비스를 운영하면서 재고가 1억개를 넘기는 거의 불가능하죠!! 이부분은 Int로 가도 될 것 같네요~!! 👍

mindock replied:
> 이부분은 내가 코틀린하면서 가장고민하는점인데, 외부에 setter를 노출하는 것과 같은 효과이자너?? 이런부분은 어떻게 생각해?

요 부분은 저도 항상 고민인데...
변경 가능성이 있다고 생각해서 `var`를 사용하게 되었는데, 외부에서 접근해서 변경할 수 있어서 위험이 있는 것 같아요. 내부에서만 변경할 수 있도록 하면 좋을 것 같은데... 요건 한번 같이 찾아봐요~!

vsh123 commented on src/main/kotlin/com/mindock/minifullfillment/service/CenterService.kt:
```diff
@@ -0,0 +1,33 @@
+        return centerRepository.save(center)
+    }
+
+    fun find(centerId: Long): Center =
+        centerRepository.findByIdOrNull(centerId)
+            ?: throw IllegalArgumentException("Center(ID: $centerId)를 찾을 수 없습니다.")
```
나는 find와 get에 대해서 차이를 두고 사용하는게 좋다고 생각해!
예를들어 find는  nullable한 값을 넘겨줄 수 있다면, get은 값이 없다면 에러를 던질 수 있다는 팀간의 약속???

그리구 자바에서는 이렇게 도메인 서비스에 find라는 함수를 만들고 다른 곳에서도 필요하면 CenterService를 호출해야했었는데, kotlin은 확장함수를 사용하니까 굳이 Service를 의존하지 않아도 좋더라구!! 내가 구현한거 [참고링크!](https://github.com/vsh123/mini-fulfillment/pull/1/files#diff-e04d0dd013c43d5e8adde9ba4198e943231add6c819eefb753bbd1811e3765b0R6)

이렇게 하니까 서비스 간의 의존이 조금은 해소되는 것 같아서, 나는 개인적으로 이러한 단순 조회성 함수는 확장함수로 사용하려고하는데, 님 생각은 어때여?

mindock replied:
궁금한게 그러면 단건 조회의 경우, controller 단에서 repository를 호출해 가져오나요?

vsh123 commented on src/main/kotlin/com/mindock/minifullfillment/service/CenterService.kt:
```diff
@@ -0,0 +1,33 @@
+    private val centerRepository: CenterRepository
+) {
+
+    fun create(request: CenterCreateRequest): Center {
```
여기에는 Transactional을 안붇이신 이유가 있나용

mindock replied:
여기서는 center 생성 및 저장하는 로직만 수행해서 @Transactional 로 묶지 않아도 되지 않나 라는 생각이었는데, 저기에 Transactional을 붙이면 어떤 이점이 있나요?

vsh123 replied:
아하!! 그렇군요!!
사실 이점은 없죠! 오히려 트랜잭션을 더 일찍열어서 DB커넥션을 오래 보유하고 있게되니, 필요시점에 트랜잭션을 열어주는게 더 좋은 방식이라고 생각합니다 👍

vsh123 commented on src/main/kotlin/com/mindock/minifullfillment/service/CenterService.kt:
```diff
@@ -0,0 +1,33 @@
+        return centerRepository.save(center)
+    }
+
+    fun find(centerId: Long): Center =
```
여기도 @Transactional(readOnly = true)가 있으면 더 좋겠꾼!

vsh123 replied:
`centerRepository.findByIdOrNull` 자체가 readOnly 트랜잭션이니까 이부분도 굳이 안붙여도 될 것 같아요!

vsh123 commented on src/main/kotlin/com/mindock/minifullfillment/dto/center/CenterResponse.kt:
```diff
@@ -0,0 +1,20 @@
+import com.mindock.minifullfillment.domain.Center
+import com.mindock.minifullfillment.domain.CenterStatus
+
+data class CenterResponse(
```
class별로 dto파일을 분리할 필요가 있을까??

mindock replied:
보통 묶어서 만들어?? 왜??

vsh123 replied:
dto클래스는 필요성이나 목적성에 따라 하나의 entity에도  n개의 클래스가 생성될 수 있다고 생각해여!

예를들어 생성 Request와 수정 Request가 각각 나올수 있는데,
자바에서는 이에 따라서 클래스 파일을 계속 생성해줘야했죠!!

서비스가 커지면 dto가 너무 많이 늘어나서 한눈에 보기힘들었는데, 이를 또 도메인별 패키지로 나누니 depth가 증가하는게 좋아보이지는 않았어요~!!

근데 코틀린에서는 1file 1class가 아니다보니 하나의 파일에 dtos라고 해서 몰아넣을 수 있어, 저는 이방식을 사용하고 있습니다. (CenterDtos.kt처럼요!)

@mindock 은 어떻게 쓰고 있어요?

mindock replied:
저는 파일마다 하나씩 만들고 있는데요~!
도메인별로 패키지를 나눠서 관리하고 있는데 아직 큰 불편은 느끼지 못해서 요 부분은 좀 더 생각해볼게요.

vsh123 commented on src/main/kotlin/com/mindock/minifullfillment/service/SkuService.kt:
```diff
@@ -0,0 +1,35 @@
+    private val skuRepository: SkuRepository,
+) {
+
+    fun create(request: SkuCreateRequest): Sku {
```
> SKU는 고유의 코드를 가지고 있습니다.

라는 요건에 대한 validation을 현재 수행해주고 있지 않은 것 같아요!

vsh123 commented on src/main/kotlin/com/mindock/minifullfillment/domain/Sku.kt:
```diff
@@ -0,0 +1,35 @@
+    @Column(name = "sku_id")
+    val id: Long = 0L,
+
+    val name: String,
```
나는 최근에는 굳이 명시가 필요하지 않은 컬럼에도 @Column어노테이션을 통해 명시를 해주려고 하고 있어!
해당 컬럼명부터, String의 경우 column의 최대길이, nullable, unique같은것들을 명시해주면 훨씬 더 보기 좋더라구!
님은 요고 어떻게 생각해여?

mindock replied:
굳 좋아요~! 👍

vsh123 reviewed (changes requested):
빠른시간에 넘나 잘해주셨쓰ㅃ니다!!!!!!!!!!!!!!

mindock committed:
� �¢�쇰�移대� import �帷굅

vsh123 commented:
코멘트는 달렸는데, 아직 수정 커밋은 push안된 것 같아요~~~~!! 확인한번 부탁드려용~~~~~~~~~~~

mindock committed:
� @Column 紐�

mindock committed:
� �⑥ �� � 以¦諛� �帷굅

mindock committed:
� sku 肄� validation 異�

- "SKU� 怨풰�� 肄�瑜 媛¢吏¢怨 ��."
  - SKU �庖¦ � ��λ��¢ sku 肄�媛¢ 議댁頡�� 泥댄¬ы�� 濡黝� 異�

mindock committed:
� Sku.of()� status 留ㅺ�蹂¢� 異�

vsh123 reviewed (approved):
고생했어요~~~~~~~!!!!!!!!
